core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java

```diff
@@ -0,0 +1,171 @@
+  protected static final String RECURSIVE_COPY = "recursive.copy";
+  protected static final int DEFAULT_MAX_SPLIT_SIZE = 128;
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractMetadataInputFormat.class);
+
+  public AbstractMetadataInputFormat() {
+  }
+
+  @Override
+  public List<InputSplit> getSplits(JobContext jobContext) throws IOException, InterruptedException {
+    // connect to the source filesystem
+    Configuration conf = jobContext.getConfiguration();
+    URI uri = URI.create(conf.get(FS_URI));
+    String[] sourcePaths = conf.get(SOURCE_PATHS).split(",");
+    Boolean recursiveCopy = conf.getBoolean(RECURSIVE_COPY, true);
+    FileSystem fileSystem = FileSystem.get(uri, conf);
+    int maxSplitSize = conf.getInt(MAX_SPLIT_SIZE, DEFAULT_MAX_SPLIT_SIZE);
+
+    List<FileStatus> fileStatuses = new ArrayList<>();
+    List<String> baseFolders = new ArrayList<>();
+    for (String prefix : sourcePaths) {
+      try {
+        FileStatus fileStatus = fileSystem.getFileStatus(new Path(prefix));
+        if (fileStatus.isFile()) {
+          prefix = "";
+        }
+        recursivelyAddFileStatus(fileStatuses, baseFolders, prefix, fileStatus, recursiveCopy, fileSystem);
+      } catch (Exception e) {
+        LOG.warn(e.getMessage());
+      }
+    }
+
+    fileSystem.close();
+
+    // create a filemetadata list of all the files
+    List<AbstractFileMetadata> fileMetaDataList = getFileMetadaDataList(fileStatuses, baseFolders,
+                                                                        fileStatuses.size(),
+                                                                        getCredentialsFromConf(conf));
+
+    // compute number of splits and instantiate the splits
+    int numSplits = (fileMetaDataList.size() - 1) / maxSplitSize + 1;
+    PriorityQueue<AbstractMetadataInputSplit> abstractInputSplits = new PriorityQueue<>(numSplits);
+    for (int i = 0; i < numSplits; i++) {
+      abstractInputSplits.add(getInputSplit());
+    }
+
+    // assign each split approximately the same number of bytes (2-approx)
+    for (int i = 0; i < fileMetaDataList.size(); i++) {
+      AbstractMetadataInputSplit minInputSplit = abstractInputSplits.poll();
+      minInputSplit.addFileMetadata(fileMetaDataList.get(i));
+      abstractInputSplits.add(minInputSplit);
+    }
+
+    // cast from AbstractMetadataInputSplit to InputSplit
+    List<InputSplit> inputSplits = new ArrayList<>();
+    inputSplits.addAll(abstractInputSplits);
+
+    return inputSplits;
+  }
+
+
+  @Override
+  public RecordReader createRecordReader(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+    throws IOException, InterruptedException {
+    MetadataRecordReader recordReader = new MetadataRecordReader();
+    recordReader.initialize(inputSplit, taskAttemptContext);
+    return recordReader;
+  }
+
+  private void recursivelyAddFileStatus (List<FileStatus> fileStatuses, List<String> baseFolders,
```

Comment:
`listFiles(Path, boolean)` ignores empty folders, which is why I wrote my own method.